firedrake/variational_solver.py

```diff
@@ -52,7 +52,7 @@ class NonlinearVariationalProblem(NonlinearVariationalProblemMixin):
     def __init__(self, F, u, bcs=None, J=None,
                  Jp=None,
                  form_compiler_parameters=None,
-                 is_linear=False, restrict=False):
+                 is_linear=False, restrict=False, pre_apply_bcs=True):
```

Comment:
I think `pre_apply_bcs` should be attached to `NonlinearVariationalSolver` instead of to `NonlinearVariationalProblem`.

Reply:
Oh I see, restrict actually changes the problem symbolically, but `pre_apply_bcs` also modifies the residual with the BC lifting. Shouldn't these two options be treated equally?

Reply:
It makes more sense to attach pre_apply_bcs to the _SNESContext, as the subproblems in multigrid and fieldsplit do not have a proper `NonlinearVariationalSolver` attached to them.